src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Reload/ProjectReloadManager.cs

```diff
@@ -0,0 +1,560 @@
+
+            if (SolutionEventsCookie == VSConstants.VSCOOKIE_NIL)
+            {
+                IVsSolution solution = _serviceProvider.GetService<IVsSolution, SVsSolution>();
+                if (solution != null)
+                {
+                    uint cookie;
+                    int hr = solution.AdviseSolutionEvents(this, out cookie);
+                    SolutionEventsCookie = cookie;
+                    System.Diagnostics.Debug.Assert(ErrorHandler.Succeeded(hr) && SolutionEventsCookie != VSConstants.VSCOOKIE_NIL);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Removes the file change watch on the project file. 
+        /// </summary>
+        private async Task DisconnectFromSolutionEvents()
+        {
+            await _threadHandling.SwitchToUIThread();
+
+            if(SolutionEventsCookie != VSConstants.VSCOOKIE_NIL)
+            {
+                IVsSolution solution = _serviceProvider.GetService<IVsSolution, SVsSolution>();
+                if (solution != null)
+                {
+                    int hr = solution.UnadviseSolutionEvents(SolutionEventsCookie);
+                    System.Diagnostics.Debug.Assert(ErrorHandler.Succeeded(hr));
+                    SolutionEventsCookie = VSConstants.VSCOOKIE_NIL;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Handles one time initialization
+        /// </summary>
+        protected override async Task InitializeCoreAsync(CancellationToken cancellationToken)
+        {
+            ReloadDelayScheduler = new TaskDelayScheduler(TimeSpan.FromMilliseconds(ReloadDelay), _threadHandling, CancellationToken.None);
+            await ConnectToSolutionEvents().ConfigureAwait(false);
+
+        }
+
+        /// <summary>
+        /// IDisposable handler. Should only be called once
+        /// </summary>
+        protected override async Task DisposeCoreAsync(bool initialized)
+        {
+            if(ReloadDelayScheduler != null)
+            {
+                ReloadDelayScheduler.Dispose();
+                ReloadDelayScheduler = null;
+            }
+            await DisconnectFromSolutionEvents().ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Called when one of the project files changes. In our case since only one file is watched with each cookie so the list of files
+        /// should be one.
+        /// </summary>
+        public int FilesChanged(uint cChanges, string[] rgpszFile, uint[] grfChange)
+        {
+            if(cChanges == 1 && (grfChange[0] & (uint)(_VSFILECHANGEFLAGS.VSFILECHG_Size | _VSFILECHANGEFLAGS.VSFILECHG_Time)) != 0)
+            {
+                IReloadableProject changedProject = null;
+                lock(_registeredProjects)
+                {
+                    changedProject = _registeredProjects.FirstOrDefault(kv => kv.Key.ProjectFile.Equals(rgpszFile[0], StringComparison.OrdinalIgnoreCase)).Key;
```

Comment:
Given this lock(_registeredProjects) pattern, I wonder if it will be a little simpler by using a ConcurrentDictionary.